resources/assets/coffee/react/ranking/page.coffee

```diff
@@ -0,0 +1,251 @@
+      page: props.paging.page
+      page_size: props.scores.length
+      pages: props.paging.pages
+      loading: false
+      mode: props.mode
+      ranking_type: 'performance' # hard coded until other types are implemented
+
+
+  changePage: (page) =>
+    @setState page: page, @retrieve
+
+
+  generateHash: =>
+    if @state.page > 0
+      laroute.route 'ranking',
+        mode: @state.mode
+        type: @state.ranking_type
+        page: (@state.page + 1)
+    else
+      laroute.route 'ranking',
+        mode: @state.mode
+        type: @state.ranking_type
+
+  updateHash: =>
+    newUrl = @generateHash()
+
+    return if newUrl == location.pathname
+
+    history.pushState history.state, null, newUrl
+
+
+  retrieve: =>
+    @setState loading: true, =>
+      $.ajax @generateHash(),
+        method: 'get'
+        dataType: 'json'
+
+      .done (data) =>
+        newState =
+          data: data.scores
+          page: data.paging.page
+          pages: data.paging.pages
+          loading: false
+
+        @setState newState, ->
+          @updateHash()
+
+
+  setCurrentPlaymode: (_e, {mode}) =>
+    return if @state.loading or @state.mode == mode
+
+    @setState mode: mode, page: 0, @retrieve
+
+
+  switchRankingTab: (_e, {tab}) =>
+    return if @state.ranking_type == tab
+
+    @setState ranking_type: tab, @retrieve
+
+
+  componentDidMount: =>
+    $.subscribe 'playmode:set.rankingPage', @setCurrentPlaymode
+    $.subscribe 'rankingmode:set.rankingPage', @setCurrentRankingMode
+    $.subscribe 'tabs:switch:ranking.rankingPage', @switchRankingTab
+    $.subscribe 'tabs:switch:sorting.rankingPage', @switchSortingTab
+
+
+  componentWillUnmount: =>
+    $.unsubscribe '.rankingPage'
+
+
+  renderRank: (props) =>
+    div className: 'ranking-page__rank-column',
+      "##{@state.page * @state.page_size + props.index + 1}"
+
+  renderUserLink: (props) ->
+    a
+      href: laroute.route 'users.show', user: props.row.user.id
+      className: 'ranking-page__user-link'
+      span
+        className: 'flag-country'
+        title: props.row.user.country_code
+        style:
+          backgroundImage: "url('/images/flags/#{props.row.user.country_code}.png')"
+      span
+        className: 'ranking-page__user-link-text'
+        props.row.user.username
+
+
+  playmodeTabHrefFunc: (mode) =>
+    laroute.route 'ranking',
+      mode: mode
+      type: @state.ranking_type
+
+
+  rankingTypeTabHrefFunc: (type) =>
+    laroute.route 'ranking',
+      mode: @state.mode
+      type: type
+
+
+  render: =>
+    # override defaults here because setting 'sortable: false' on
+    #   the table itself doesn't disable sorting, idk
+    ReactTable.ReactTableDefaults.column.sortable = false
+
+    div null,
+      div className: 'osu-page',
+        el PlaymodeTabs,
+          enableAll: true
+          currentMode: @state.mode
+          hrefFunc: @playmodeTabHrefFunc
+
+        div
+          className: 'ranking-page__header'
```

Comment:
not only the block is missing but it looks like you're turning two different blocks into elements <_<